src/rust/engine/src/externs/fs.rs

```diff
@@ -199,6 +207,68 @@ impl PySnapshot {
   }
 }
 
+#[pyclass]
+#[derive(Clone, Debug, PartialEq)]
+pub struct PyAddPrefix(pub Digest, pub PathBuf);
+
+#[pymethods]
+impl PyAddPrefix {
+  #[new]
+  fn __new__(digest: PyDigest, prefix: PathBuf) -> Self {
+    Self(digest.0, prefix)
+  }
+
+  fn __hash__(&self) -> u64 {
+    let mut s = DefaultHasher::new();
+    self.0.hash.prefix_hash().hash(&mut s);
+    self.1.hash(&mut s);
+    s.finish()
+  }
+
+  fn __repr__(&self) -> String {
+    format!("AddPrefix('{}', {})", PyDigest(self.0), self.1.display())
+  }
+
+  fn __richcmp__(&self, other: &PyAddPrefix, op: CompareOp, py: Python) -> PyObject {
+    match op {
+      CompareOp::Eq => (self == other).into_py(py),
+      CompareOp::Ne => (self != other).into_py(py),
+      _ => py.NotImplemented(),
+    }
+  }
+}
+
+#[pyclass]
+#[derive(Clone, Debug, PartialEq)]
+pub struct PyRemovePrefix(pub Digest, pub PathBuf);
```

Comment:
Why not use a struct instead of a tuple? Having names for the `Digest` and `PathBuf` would be good for future readers of the code.

Reply:
Same comment for `PyAddPrefix`